File: t/CS/SVGDemo/UI/Viewer.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;

using Leadtools;
using Leadtools.Svg;
using Leadtools.Drawing;
using System.Collections.Generic;
using System.Drawing.Drawing2D;

namespace SvgDemo
{
   public class Viewer : Control
   {
      public Viewer()
      {
         var styles = ControlStyles.AllPaintingInWmPaint | ControlStyles.OptimizedDoubleBuffer | ControlStyles.ResizeRedraw | ControlStyles.UserPaint;
         SetStyle(styles, true);
         _currentPageIndex = 0;
         _totalPages = 1;
      }

      private BorderStyle _borderStyle = BorderStyle.None;
      [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
      public BorderStyle BorderStyle
      {
         get
         {
            return _borderStyle;
         }
         set
         {
            if (_borderStyle != value)
            {
               _borderStyle = value;
               UpdateStyles();
               Invalidate();
            }
         }
      }

      // To support border styles
      protected override CreateParams CreateParams
      {
         get
         {
            CreateParams cp = base.CreateParams;

            switch (BorderStyle)
            {
               case BorderStyle.None:
                  cp.Style &= ~SafeNativeMethods.WS_BORDER;
                  cp.ExStyle &= ~SafeNativeMethods.WS_EX_CLIENTEDGE;
                  break;

               case BorderStyle.FixedSingle:
                  cp.Style |= SafeNativeMethods.WS_BORDER;
                  cp.ExStyle &= ~SafeNativeMethods.WS_EX_CLIENTEDGE;
                  break;

               case BorderStyle.Fixed3D:
                  cp.Style &= ~SafeNativeMethods.WS_BORDER;
                  cp.ExStyle |= SafeNativeMethods.WS_EX_CLIENTEDGE;
                  break;
            }

            return cp;
         }
      }

      private LeadMatrix _transformInverse = LeadMatrix.Identity;
      private LeadMatrix _transform = LeadMatrix.Identity;
      [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
      public LeadMatrix Transform
      {
         get { return _transform; }
         set
         {
            _transform = value;
            _transformInverse = _transform.Clone();
            if (_transformInverse.HasInverse)
               _transformInverse.Invert();

            Invalidate();
         }
      }

      private List<SvgDocumentInformation> _documents = new List<SvgDocumentInformation>();
      public List<SvgDocumentInformation> DocumentList
      {
         get
         {
            return _documents;
         }

         set
         {
            foreach (SvgDocumentInformation doc in _documents)
            {
               doc.Document.Dispose();
            }

            _documents.Clear();
            _currentPageIndex = 0;
            _totalPages = 0;

            if (value != null)
            {
               foreach (SvgDocumentInformation doc in value)
                  _documents.Add(doc);
            }
         }
      }

      [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
      public SvgDocument CurrentDocument
      {
         get
         {
            if (_documents.Count == 0)
               return null;

            return _documents[_currentPageIndex].Document;
         }
      }

      public enum CoordinateType
      {
         Image,
         Control
      }

      public LeadPointD ConvertPoint(CoordinateType source, CoordinateType destination, LeadPointD point)
      {
         if (source == destination)
            return point;

         LeadPointD[] points = { point };

         if (source == CoordinateType.Control)
            _transformInverse.TransformPoints(points);
         else
            _transform.TransformPoints(points);

         return points[0];
      }

      private bool _isPanning = false;

      private void StopPanning()
      {
         if (_documents.Count != 0 && _isPanning)
         {
            this.Capture = false;
            _isPanning = false;

            if (PanEnd != null)
               PanEnd(this, EventArgs.Empty);
         }
      }

      private void StartPanning(MouseEventArgs e)
      {
         if (_documents.Count == 0)
            return;

         _isPanning = true;
         this.Capture = true;

         if (PanBegin != null)
            PanBegin(this, e);
      }

      protected override void OnMouseDown(MouseEventArgs e)
      {
         StopPanning();

         Focus();
         StartPanning(e);

         base.OnMouseDown(e);
      }

      public event MouseEventHandler Panning;
      public event MouseEventHandler PanBegin;
      public event EventHandler PanEnd;

      protected override void OnMouseMove(MouseEventArgs e)
      {
         if (_documents.Count != 0 && _isPanning)
         {
            if (Panning != null)
               Panning(this, e);
         }

         base.OnMouseMove(e);
      }

      protected override void OnMouseUp(MouseEventArgs e)
      {
         StopPanning();

         base.OnMouseUp(e);
      }

      protected override void OnLostFocus(EventArgs e)
      {
         StopPanning();

         base.OnLostFocus(e);
      }

      private bool _useDpi;

      public bool UseDpi
      {
         get { return _useDpi; }
         set { _useDpi = value; Identity(); }
      }

      public LeadSize ImageSize
      {
         get
         {
            if (_documents[_currentPageIndex].Document == null)
               return LeadSize.Create(0, 0);

            var svgBounds = _documents[_currentPageIndex].Document.Bounds;
            return LeadSize.Create(Tools.ToInt(svgBounds.Bounds.Width), Tools.ToInt(svgBounds.Bounds.Height));
         }
      }

      public void Identity()
      {
         if (_documents.Count == 0)
            return;

         if (_documents[_currentPageIndex].Document == null)
            return;

         var svgBounds = _documents[_currentPageIndex].Document.Bounds;
         var bounds = svgBounds.Bounds;
         var transform = LeadMatrix.Identity;

         if (_useDpi)
         {
            var imageSize = this.ImageSize;
            var resolution = LeadSizeD.Create(96, 96);
            if (!svgBounds.Resolution.IsEmpty)
            {
               if (svgBounds.Resolution.Width > 0)
                  resolution.Width = svgBounds.Resolution.Width;
               if (svgBounds.Resolution.Height > 0)
                  resolution.Height = svgBounds.Resolution.Height;
            }

            transform.Scale(
               (double)96.0 / resolution.Width,
               (double)96.0 / resolution.Height);
         }

         this.Transform = transform;
         Invalidate();
      }

      private bool _drawClipRectangle = true;
      public bool DrawClipRectangle
      {
         get { return _drawClipRectangle; }
         set
         {
            _drawClipRectangle = value;
            Invalidate();
         }
      }

      private int _updateCounter = 0;
      public void BeginUpdate()
      {
         _updateCounter++;
      }

      public void EndUpdate()
      {
         if (_updateCounter == 0) throw new InvalidOperationException("EndUpdate without matching BeginUpdate");
         _updateCounter--;

         if (_updateCounter == 0)
            Invalidate();
      }

      protected bool CanUpdate
      {
         get { return _updateCounter == 0; }
      }

      protected override void OnPaint(PaintEventArgs e)
      {
         DrawImage(e, this.ClientRectangle);
         base.OnPaint(e);
      }

      private static void DrawBounds(Graphics graphics, Pen pen, Brush brush, Brush textBrush, Font font, string text, LeadRectD bounds, LeadMatrix transform)
      {
         var corners = Tools.GetCornerPoints(bounds);
         transform.TransformPoints(corners);
         var cornersBounds = Tools.GetBoundingRect(corners);

         var cornersF = new PointF[4];
         for (var i = 0; i < cornersF.Length; i++)
         {
            // if any point empty, then return
            if (corners[i].IsEmpty)
               return;

            cornersF[i] = new PointF((float)corners[i].X, (float)corners[i].Y);
         }

         if (pen != null)
            graphics.DrawPolygon(pen, cornersF);
         if (brush != null)
            graphics.FillPolygon(brush, cornersF);
         if (text != null && textBrush != null)
            graphics.DrawString(text, font, textBrush, (float)cornersBounds.X, (float)cornersBounds.Y);
      }

      private void DrawImage(PaintEventArgs e, Rectangle clipRect)
      {
         if (_documents.Count == 0)
            return;

         if (_documents[_currentPageIndex].Document == null || !this.CanUpdate)
            return;

         var graphics = e.Graphics;

         var options = new SvgRenderOptions();
         options.Transform = _transform;
         options.Bounds = _documents[_currentPageIndex].Document.Bounds.Bounds;
         options.UseBackgroundColor = true;
         options.ClipBounds = LeadRectD.Create(clipRect.X, clipRect.Y, clipRect.Width, clipRect.Height);
         options.BackgroundColor = RasterColor.FromKnownColor(RasterKnownColor.White);

         try
         {
            using (var engine = RenderingEngineFactory.Create(graphics))
               _documents[_currentPageIndex].Document.Render(engine, options);
         }
         catch
         {
            Console.WriteLine();
         }

         DrawBounds(graphics, Pens.Black, null, null, null, null, options.Bounds, options.Transform);

         if (_documents[_currentPageIndex].DocumentText != null && _documents[_currentPageIndex].ShowText)
         {
            LeadRectD docBounds = _documents[_currentPageIndex].Document.Bounds.Bounds;

            // Could be rotated, so
            LeadPointD topLeft = docBounds.TopLeft;
            LeadPointD bottomRight = docBounds.BottomRight;

            LeadPointD[] corners = new LeadPointD[4];
            corners[0].X = topLeft.X;
            corners[0].Y = topLeft.Y;
            corners[1].X = bottomRight.X;
            corners[1].Y = topLeft.Y;
            corners[2].X = bottomRight.X;
            corners[2].Y = bottomRight.Y;
            corners[3].X = topLeft.X;
            corners[3].Y = bottomRight.Y;

            options.Transform.TransformPoints(corners);

            GraphicsPath path = new GraphicsPath();
            PointF[] pts = new PointF[4];
            for (int i = 0; i < corners.Length; i++)
            {
               pts[i].X = (float)corners[i].X;
               pts[i].Y = (float)corners[i].Y;
            }
            path.AddPolygon(pts);
            graphics.SetClip(path, System.Drawing.Drawing2D.CombineMode.Intersect);

            using (var brush = new SolidBrush(Color.FromArgb(64, Color.Black)))
            {
               foreach (var character in _documents[_currentPageIndex].DocumentText.Characters)
               {
                  var bounds = character.Bounds;
                  var text = new string(new char[] { character.Code });

                  DrawBounds(graphics, Pens.Yellow, brush, Brushes.Yellow, Font, text, character.Bounds, options.Transform);
               }
            }
         }

         base.OnPaint(e);
      }

      private int _currentPageIndex;
      public int CurrentPageIndex
      {
         get { return _currentPageIndex; }
         set
         {
            _currentPageIndex = value;

            Identity();
            Invalidate();
         }
      }

      private int _totalPages;
      public int TotalPages
      {
         get { return _totalPages; }
         set { _totalPages = value; }
      }
   }
}

File: t/VB/SVGDemo/UI/Viewer.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************

Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms

Imports Leadtools
Imports Leadtools.Svg
Imports Leadtools.Drawing
Imports System.Collections.Generic
Imports System.Drawing.Drawing2D
Imports System

Namespace SvgDemo
   Public Class Viewer
      Inherits Control
      Public Sub New()
         Dim styles As ControlStyles = ControlStyles.AllPaintingInWmPaint Or ControlStyles.OptimizedDoubleBuffer Or ControlStyles.ResizeRedraw Or ControlStyles.UserPaint
         SetStyle(styles, True)
         _currentPageIndex = 0
         _totalPages = 1
      End Sub

      Private _borderStyle As BorderStyle = BorderStyle.None
      <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
      Public Property BorderStyle() As BorderStyle
         Get
            Return _borderStyle
         End Get
         Set(value As BorderStyle)
            If _borderStyle <> value Then
               _borderStyle = value
               UpdateStyles()
               Invalidate()
            End If
         End Set
      End Property

      ' To support border styles
      Protected Overrides ReadOnly Property CreateParams() As CreateParams
         Get
            Dim cp As CreateParams = MyBase.CreateParams

            Select Case BorderStyle
               Case BorderStyle.None
                  cp.Style = cp.Style And Not SafeNativeMethods.WS_BORDER
                  cp.ExStyle = cp.ExStyle And Not SafeNativeMethods.WS_EX_CLIENTEDGE
                  Exit Select

               Case BorderStyle.FixedSingle
                  cp.Style = cp.Style Or SafeNativeMethods.WS_BORDER
                  cp.ExStyle = cp.ExStyle And Not SafeNativeMethods.WS_EX_CLIENTEDGE
                  Exit Select

               Case BorderStyle.Fixed3D
                  cp.Style = cp.Style And Not SafeNativeMethods.WS_BORDER
                  cp.ExStyle = cp.ExStyle Or SafeNativeMethods.WS_EX_CLIENTEDGE
                  Exit Select
            End Select

            Return cp
         End Get
      End Property

      Private _transformInverse As LeadMatrix = LeadMatrix.Identity
      Private _transform As LeadMatrix = LeadMatrix.Identity
      <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
      Public Property Transform() As LeadMatrix
         Get
            Return _transform
         End Get
         Set(value As LeadMatrix)
            _transform = value
            _transformInverse = _transform.Clone()
            If _transformInverse.HasInverse Then
               _transformInverse.Invert()
            End If

            Invalidate()
         End Set
      End Property

      Private _documents As New List(Of SvgDocumentInformation)()
      Public Property DocumentList() As List(Of SvgDocumentInformation)
         Get
            Return _documents
         End Get

         Set(value As List(Of SvgDocumentInformation))
            For Each doc As SvgDocumentInformation In _documents
               doc.Document.Dispose()
            Next

            _documents.Clear()
            _currentPageIndex = 0
            _totalPages = 0

            If value IsNot Nothing Then
               For Each doc As SvgDocumentInformation In value
                  _documents.Add(doc)
               Next
            End If
         End Set
      End Property

      <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
      Public ReadOnly Property CurrentDocument() As SvgDocument
         Get
            If _documents.Count = 0 Then
               Return Nothing
            End If

            Return _documents(_currentPageIndex).Document
         End Get
      End Property

      Public Enum CoordinateType
         Image
         Control
      End Enum

      Public Function ConvertPoint(source As CoordinateType, destination As CoordinateType, point As LeadPointD) As LeadPointD
         If source = destination Then
            Return point
         End If

         Dim points As LeadPointD() = {point}

         If source = CoordinateType.Control Then
            _transformInverse.TransformPoints(points)
         Else
            _transform.TransformPoints(points)
         End If

         Return points(0)
      End Function

      Private _isPanning As Boolean = False

      Private Sub StopPanning()
         If _documents.Count <> 0 AndAlso _isPanning Then
            Me.Capture = False
            _isPanning = False

            RaiseEvent PanEnd(Me, EventArgs.Empty)
         End If
      End Sub

      Private Sub StartPanning(e As MouseEventArgs)
         If _documents.Count = 0 Then
            Return
         End If

         _isPanning = True
         Me.Capture = True

         RaiseEvent PanBegin(Me, e)
      End Sub

      Protected Overrides Sub OnMouseDown(e As MouseEventArgs)
         StopPanning()

         Focus()
         StartPanning(e)

         MyBase.OnMouseDown(e)
      End Sub

      Public Event Panning As MouseEventHandler
      Public Event PanBegin As MouseEventHandler
      Public Event PanEnd As EventHandler

      Protected Overrides Sub OnMouseMove(e As MouseEventArgs)
         If _documents.Count <> 0 AndAlso _isPanning Then
            RaiseEvent Panning(Me, e)
         End If

         MyBase.OnMouseMove(e)
      End Sub

      Protected Overrides Sub OnMouseUp(e As MouseEventArgs)
         StopPanning()

         MyBase.OnMouseUp(e)
      End Sub

      Protected Overrides Sub OnLostFocus(e As EventArgs)
         StopPanning()

         MyBase.OnLostFocus(e)
      End Sub

      Private _useDpi As Boolean

      Public Property UseDpi() As Boolean
         Get
            Return _useDpi
         End Get
         Set(value As Boolean)
            _useDpi = value
            Identity()
         End Set
      End Property

      Public ReadOnly Property ImageSize() As LeadSize
         Get
            If _documents(_currentPageIndex).Document Is Nothing Then
               Return LeadSize.Create(0, 0)
            End If

            Dim svgBounds As SvgBounds = _documents(_currentPageIndex).Document.Bounds



            Return LeadSize.Create(Tools.ToInt(svgBounds.Bounds.Width), Tools.ToInt(svgBounds.Bounds.Height))
         End Get
      End Property

      Public Sub Identity()
         If _documents.Count = 0 Then
            Return
         End If

         If _documents(_currentPageIndex).Document Is Nothing Then
            Return
         End If

         Dim svgBounds As SvgBounds = _documents(_currentPageIndex).Document.Bounds



         Dim bounds As LeadRectD = svgBounds.Bounds
         Dim transform As LeadMatrix = LeadMatrix.Identity

         If _useDpi Then
            Dim imageSize As LeadSize = Me.ImageSize
            Dim resolution As LeadSizeD = LeadSizeD.Create(96, 96)
            If Not svgBounds.Resolution.IsEmpty Then
               If svgBounds.Resolution.Width > 0 Then
                  resolution.Width = svgBounds.Resolution.Width
               End If
               If svgBounds.Resolution.Height > 0 Then
                  resolution.Height = svgBounds.Resolution.Height
               End If
            End If

            transform.Scale(CDbl(96.0) / resolution.Width, CDbl(96.0) / resolution.Height)
         End If

         Me.Transform = transform
         Invalidate()
      End Sub

      Private _drawClipRectangle As Boolean = True
      Public Property DrawClipRectangle() As Boolean
         Get
            Return _drawClipRectangle
         End Get
         Set(value As Boolean)
            _drawClipRectangle = value
            Invalidate()
         End Set
      End Property

      Private _updateCounter As Integer = 0
      Public Sub BeginUpdate()
         _updateCounter += 1
      End Sub

      Public Sub EndUpdate()
         If _updateCounter = 0 Then
            Throw New InvalidOperationException("EndUpdate without matching BeginUpdate")
         End If
         _updateCounter -= 1

         If _updateCounter = 0 Then
            Invalidate()
         End If
      End Sub

      Protected ReadOnly Property CanUpdate() As Boolean
         Get
            Return _updateCounter = 0
         End Get
      End Property

      Protected Overrides Sub OnPaint(e As PaintEventArgs)
         DrawImage(e, Me.ClientRectangle)
         MyBase.OnPaint(e)
      End Sub

      Private Shared Sub DrawBounds(graphics As Graphics, pen As Pen, brush As Brush, textBrush As Brush, font As Font, text As String, _
       bounds As LeadRectD, transform As LeadMatrix)
         Dim corners As LeadPointD() = Tools.GetCornerPoints(bounds)
         transform.TransformPoints(corners)
         Dim cornersBounds As LeadRectD = Tools.GetBoundingRect(corners)

         Dim cornersF As PointF() = New PointF(3) {}
         For i As Integer = 0 To cornersF.Length - 1
            ' if any point empty, then return
            If corners(i).IsEmpty Then
               Return
            End If

            cornersF(i) = New PointF(CSng(corners(i).X), CSng(corners(i).Y))
         Next

         If pen IsNot Nothing Then
            graphics.DrawPolygon(pen, cornersF)
         End If
         If brush IsNot Nothing Then
            graphics.FillPolygon(brush, cornersF)
         End If
         If text IsNot Nothing AndAlso textBrush IsNot Nothing Then
            graphics.DrawString(text, font, textBrush, CSng(cornersBounds.X), CSng(cornersBounds.Y))
         End If
      End Sub

      Private Sub DrawImage(e As PaintEventArgs, clipRect As Rectangle)
         If _documents.Count = 0 Then
            Return
         End If

         If _documents(_currentPageIndex).Document Is Nothing OrElse Not Me.CanUpdate Then
            Return
         End If

         Dim graphics As Graphics = e.Graphics

         Dim options As SvgRenderOptions = New SvgRenderOptions()
         options.Transform = _transform
         options.Bounds = _documents(_currentPageIndex).Document.Bounds.Bounds
         options.UseBackgroundColor = True
         options.ClipBounds = LeadRectD.Create(clipRect.X, clipRect.Y, clipRect.Width, clipRect.Height)
         options.BackgroundColor = RasterColor.FromKnownColor(RasterKnownColor.White)

         Try
            Using engine As IRenderingEngine = RenderingEngineFactory.Create(graphics)
               _documents(_currentPageIndex).Document.Render(engine, options)
            End Using
         Catch
            Console.WriteLine()
         End Try

         DrawBounds(graphics, Pens.Black, Nothing, Nothing, Nothing, Nothing, _
          options.Bounds, options.Transform)

         If _documents(_currentPageIndex).DocumentText IsNot Nothing AndAlso _documents(_currentPageIndex).ShowText Then
            Dim docBounds As LeadRectD = _documents(_currentPageIndex).Document.Bounds.Bounds

            ' Could be rotated, so
            Dim topLeft As LeadPointD = docBounds.TopLeft
            Dim bottomRight As LeadPointD = docBounds.BottomRight

            Dim corners As LeadPointD() = New LeadPointD(3) {}
            corners(0).X = topLeft.X
            corners(0).Y = topLeft.Y
            corners(1).X = bottomRight.X
            corners(1).Y = topLeft.Y
            corners(2).X = bottomRight.X
            corners(2).Y = bottomRight.Y
            corners(3).X = topLeft.X
            corners(3).Y = bottomRight.Y

            options.Transform.TransformPoints(corners)

            Dim path As New GraphicsPath()
            Dim pts As PointF() = New PointF(3) {}
            For i As Integer = 0 To corners.Length - 1
               pts(i).X = CSng(corners(i).X)
               pts(i).Y = CSng(corners(i).Y)
            Next
            path.AddPolygon(pts)
            graphics.SetClip(path, System.Drawing.Drawing2D.CombineMode.Intersect)

            Using brush As SolidBrush = New SolidBrush(Color.FromArgb(64, Color.Black))
               For Each character As DocumentCharacter In _documents(_currentPageIndex).DocumentText.Characters
                  Dim bounds As LeadRectD = character.Bounds
                  Dim text As String = New String(New Char() {character.Code})

                  DrawBounds(graphics, Pens.Yellow, brush, Brushes.Yellow, Font, text, _
                   character.Bounds, options.Transform)
               Next
            End Using
         End If

         MyBase.OnPaint(e)
      End Sub

      Private _currentPageIndex As Integer
      Public Property CurrentPageIndex() As Integer
         Get
            Return _currentPageIndex
         End Get
         Set(value As Integer)
            _currentPageIndex = value




            Identity()
            Invalidate()
         End Set
      End Property

      Private _totalPages As Integer
      Public Property TotalPages() As Integer
         Get
            Return _totalPages
         End Get
         Set(value As Integer)
            _totalPages = value
         End Set
      End Property
   End Class
End Namespace
